File: emulator/src/emulator/opcodes/move.py
from more_itertools import flatten

from emulator.adressing import Immediate, ZeroPage, Absolute, ZeroPageY, AbsoluteY, IndirectY, ZeroPageX, AbsoluteX, IndirectX
from emulator.constants import NEGATIVE_BIT, LOW_BITS_MASK
from emulator.opcodes.base import OpCode


class LDA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xA1, IndirectX, 6),
                      (0xA5, ZeroPage, 3,),
                      (0xA9, Immediate, 2,),
                      (0xAD, Absolute, 4,),
                      (0xB1, IndirectY, 5),
                      (0xB5, ZeroPageX, 4,),
                      (0xB9, AbsoluteY, 4,),
                      (0xBD, AbsoluteX, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_lda():
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                value = self.addressing_mode.read_from(cpu, memory, address,ld=1)
                if self.addressing_mode != Immediate:
                    self.addressing_mode.data = "= %02X" % memory.fetch(address)
                    cpu.addr = memory.get_effective_address(address)
                    cpu.data = value

                cpu.a = value
                cpu.zero = cpu.a == 0
                cpu.negative = (cpu.a & 0b10000000) > 0

        cycle_lda()


class STA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x81, IndirectX, 6,),
                      (0x85, ZeroPage, 3,),
                      (0x8D, Absolute, 4,),
                      (0x91, IndirectY, 6,),
                      (0x95, ZeroPageX, 4,),
                      (0x99, AbsoluteY, 5,),
                      (0x9D, AbsoluteX, 5,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def _stall():
            pass

        def cycle_sta():
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                cpu.addr = memory.get_effective_address(address)
                cpu.data = cpu.a
                self.addressing_mode.data = "= %02X" % memory.fetch(address)
                self.addressing_mode.write_to(cpu, memory, address, cpu.a)

        if self.addressing_mode in [IndirectY, AbsoluteY, AbsoluteX]:
            # FIXME: this is ugly, but it works
            cycle_start = cpu.cycle
            cycle_sta()
            cycle_end = cpu.cycle
            if (cycle_end - cycle_start) < (self.cycles - 1):
                cpu.exec_in_cycle(_stall)
        else:
            cycle_sta()


class LDX(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xA2, Immediate, 2,),
                      (0xA6, ZeroPage, 3,),
                      (0xAE, Absolute, 4,),
                      (0xB6, ZeroPageY, 4,),
                      (0xBE, AbsoluteY, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_ldx():
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                value = self.addressing_mode.read_from(cpu, memory, address,ld=1)
                if self.addressing_mode != Immediate:
                    self.addressing_mode.data = "= %02X" % memory.fetch(address)
                    cpu.addr = memory.get_effective_address(address)
                    cpu.data = value

                cpu.x = value
                cpu.zero = cpu.x == 0
                cpu.negative = (cpu.x & 0b10000000) > 0

        cycle_ldx()


class STX(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x86, ZeroPage, 3,),
                      (0x8E, Absolute, 4,),
                      (0x96, ZeroPageY, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_stx():
            """
            #  address R/W description
           --- ------- --- ------------------------------------------
            1    PC     R  fetch opcode, increment PC
            2    PC     R  fetch low byte of address, increment PC
            3    PC     R  fetch high byte of address, increment PC
            4  address  W  write register to effective address
            """
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                cpu.addr = memory.get_effective_address(address)
                cpu.data = cpu.x
                self.addressing_mode.data = "= %02X" % memory.fetch(address)
                self.addressing_mode.write_to(cpu, memory, address, cpu.x)

        cycle_stx()


class LDY(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xA0, Immediate, 2,),
                      (0xA4, ZeroPage, 3,),
                      (0xAC, Absolute, 4,),
                      (0xB4, ZeroPageX, 4,),
                      (0xBC, AbsoluteX, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_ldy():
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                value = self.addressing_mode.read_from(cpu, memory, address,ld=1)
                if self.addressing_mode != Immediate:
                    self.addressing_mode.data = "= %02X" % memory.fetch(address)
                    cpu.addr = memory.get_effective_address(address)
                    cpu.data = value

                cpu.y = value
                cpu.zero = cpu.y == 0
                cpu.negative = (cpu.y & 0b10000000) > 0

        cycle_ldy()


class STY(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x84, ZeroPage, 3,),
                      (0x8C, Absolute, 4,),
                      (0x94, ZeroPageX, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_sty():
            if self.addressing_mode:
                address = self.addressing_mode.fetch_address(cpu, memory)
                cpu.addr = memory.get_effective_address(address)
                cpu.data = cpu.y
                self.addressing_mode.data = "= %02X" % memory.fetch(address)
                self.addressing_mode.write_to(cpu, memory, address, cpu.y)

        cycle_sty()


class TAX(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xAA, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_tax():
            cpu.x = cpu.a
            cpu.zero = cpu.x == 0
            cpu.negative = (cpu.x & 0b10000000) > 0

        cpu.exec_in_cycle(cycle_tax)


class TXA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x8A, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_txa():
            cpu.a = cpu.x
            cpu.zero = cpu.a == 0
            cpu.negative = (cpu.a & 0b10000000) > 0

        cpu.exec_in_cycle(cycle_txa)


class TAY(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xA8, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_tay():
            cpu.y = cpu.a
            cpu.zero = cpu.y == 0
            cpu.negative = (cpu.y & 0b10000000) > 0

        cpu.exec_in_cycle(cycle_tay)


class TYA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x98, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        def cycle_tya():
            cpu.a = cpu.y
            cpu.zero = cpu.a == 0
            cpu.negative = (cpu.a & 0b10000000) > 0

        cpu.exec_in_cycle(cycle_tya)


class TSX(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0xBA, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        cpu.x = (cpu.sp & LOW_BITS_MASK)
        cpu.zero = cpu.x == 0
        cpu.negative = (cpu.x & 0b10000000) > 0
        cpu.inc_cycle()


class TXS(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x9A, None, 2,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        cpu.sp = 0x0100 | (cpu.x & LOW_BITS_MASK)
        cpu.inc_cycle()


class PLA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x68, None, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        cpu.sp += 1
        cpu.sp = cpu.sp & 0xff ^ 0x0100
        from_stack = memory.fetch(cpu.sp)
        cpu.a = from_stack
        cpu.zero = (cpu.a == 0)
        cpu.negative = (cpu.a & NEGATIVE_BIT) > 0
        cpu.inc_cycle()
        cpu.inc_cycle()
        cpu.inc_cycle()


class PHA(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x48, None, 3,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        memory.store(cpu.sp, cpu.a)
        cpu.sp -= 1
        cpu.sp = cpu.sp & 0xff ^ 0x0100
        cpu.inc_cycle()
        cpu.inc_cycle()


class PLP(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x28, None, 4,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        cpu.sp += 1
        cpu.sp = cpu.sp & 0xff ^ 0x0100
        from_stack = memory.fetch(cpu.sp)
        status = (from_stack & 0b11101111) | 0b00100000
        cpu.flags = status
        cpu.inc_cycle()
        cpu.inc_cycle()
        cpu.inc_cycle()


class PHP(OpCode):
    @classmethod
    def create_variations(cls):
        variations = [(0x08, None, 3,)]
        return map(cls.create_dict_entry, variations)

    def exec(self, cpu, memory):
        """
        In the byte pushed, bit 5 is always set to 1, and bit 4 is 1 if from an instruction (PHP or BRK) or 0 if from an interrupt line being pulled low (/IRQ or /NMI).
        """
        status = cpu.flags | 0b00110000
        memory.store(cpu.sp, status)
        cpu.sp -= 1
        cpu.sp = cpu.sp & 0xff ^ 0x0100
        cpu.inc_cycle()
        cpu.inc_cycle()


class MoveOpCodes:
    opcodes = [
        LDA,
        STA,
        LDX,
        STX,
        LDY,
        STY,
        TAX,
        TXA,
        TAY,
        TYA,
        TSX,
        TXS,
        PLA,
        PHA,
        PLP,
        PHP
    ]

    @staticmethod
    def all_commands():
        return flatten(
            map(lambda x: list(x.create_variations()), MoveOpCodes.opcodes)
        )

File: emulator-go/opcodes/move.go
package opcodes

import (
	"students.ic.unicamp.br/goten/processor"
)

type lda struct{}

func (o lda) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)
		value := byte(variation.addressingMode.ReadFrom(console, address))

		/// log
		state.HasData = true
		state.Data = value

		console.CPU.A = value
		console.CPU.SetZN(value)
	}

	if stall {
		cycleAcc++
	}

	return variation.cycles + cycleAcc
}

func (o lda) getVariations() []Variation {
	return []Variation{
		{opcode: 0xA1, addressingMode: IndirectX, cycles: 6},
		{opcode: 0xA5, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0xA9, addressingMode: Immediate, cycles: 2},
		{opcode: 0xAD, addressingMode: Absolute, cycles: 4},
		{opcode: 0xB1, addressingMode: IndirectY, cycles: 5},
		{opcode: 0xB5, addressingMode: ZeroPageX, cycles: 4},
		{opcode: 0xB9, addressingMode: AbsoluteY, cycles: 4},
		{opcode: 0xBd, addressingMode: AbsoluteX, cycles: 4},
	}
}

func (o lda) GetName() string {
	return "LDA"
}

type sta struct{}

func (o sta) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)

		/// log
		state.HasData = true
		state.Data = console.FetchData(address)

		variation.addressingMode.WriteTo(console, address, console.CPU.A)
	}

	if stall {
		cycleAcc++
	}

	// uhh, write instructions don't stall???
	return variation.cycles
}

func (o sta) getVariations() []Variation {
	return []Variation{
		{opcode: 0x81, addressingMode: IndirectX, cycles: 6},
		{opcode: 0x85, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0x8D, addressingMode: Absolute, cycles: 4},
		{opcode: 0x91, addressingMode: IndirectY, cycles: 6},
		{opcode: 0x95, addressingMode: ZeroPageX, cycles: 4},
		{opcode: 0x99, addressingMode: AbsoluteY, cycles: 5},
		{opcode: 0x9D, addressingMode: AbsoluteX, cycles: 5},
	}
}

func (o sta) GetName() string {
	return "STA"
}

type ldx struct{}

func (o ldx) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)
		value := byte(variation.addressingMode.ReadFrom(console, address))

		/// log
		state.HasData = true
		state.Data = value

		console.CPU.X = value
		console.CPU.SetZN(value)
	}

	if stall {
		cycleAcc++
	}

	return variation.cycles + cycleAcc
}

func (o ldx) getVariations() []Variation {
	return []Variation{
		{opcode: 0xA2, addressingMode: Immediate, cycles: 2},
		{opcode: 0xA6, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0xAE, addressingMode: Absolute, cycles: 4},
		{opcode: 0xB6, addressingMode: ZeroPageY, cycles: 4},
		{opcode: 0xBE, addressingMode: AbsoluteY, cycles: 4},
	}
}

func (o ldx) GetName() string {
	return "LDX"
}

type stx struct{}

func (o stx) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)

		/// log
		state.HasData = true
		state.Data = console.FetchData(address)

		variation.addressingMode.WriteTo(console, address, console.CPU.X)
	}

	if stall {
		cycleAcc++
	}

	// uhh, write instructions don't stall???
	return variation.cycles
}

func (o stx) getVariations() []Variation {
	return []Variation{
		{opcode: 0x86, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0x8E, addressingMode: Absolute, cycles: 4},
		{opcode: 0x96, addressingMode: ZeroPageY, cycles: 4},
	}
}

func (o stx) GetName() string {
	return "STX"
}

type ldy struct{}

func (o ldy) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)
		value := byte(variation.addressingMode.ReadFrom(console, address))

		/// log
		state.HasData = true
		state.Data = value

		console.CPU.Y = value
		console.CPU.SetZN(value)
	}

	if stall {
		cycleAcc++
	}

	return variation.cycles + cycleAcc
}

func (o ldy) getVariations() []Variation {
	return []Variation{
		{opcode: 0xA0, addressingMode: Immediate, cycles: 2},
		{opcode: 0xA4, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0xAC, addressingMode: Absolute, cycles: 4},
		{opcode: 0xB4, addressingMode: ZeroPageX, cycles: 4},
		{opcode: 0xBC, addressingMode: AbsoluteX, cycles: 4},
	}
}

func (o ldy) GetName() string {
	return "LDY"
}

type sty struct{}

func (o sty) Exec(console *processor.Console, variation *Variation, state *State) int {
	var stall bool = false
	var cycleAcc int = 0
	if variation.addressingMode != nil {
		var address uint16
		address, stall = variation.addressingMode.FetchAddress(console, state)

		/// log
		state.HasData = true
		state.Data = console.FetchData(address)

		variation.addressingMode.WriteTo(console, address, console.CPU.Y)
	}

	if stall {
		cycleAcc++
	}

	// uhh, write instructions don't stall???
	return variation.cycles
}

func (o sty) getVariations() []Variation {
	return []Variation{
		{opcode: 0x84, addressingMode: ZeroPage, cycles: 3},
		{opcode: 0x8C, addressingMode: Absolute, cycles: 4},
		{opcode: 0x94, addressingMode: ZeroPageX, cycles: 4},
	}
}

func (o sty) GetName() string {
	return "STY"
}

type tax struct{}

func (o tax) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.CPU.A
	console.CPU.X = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o tax) getVariations() []Variation {
	return []Variation{
		{opcode: 0xAA, addressingMode: nil, cycles: 2},
	}
}

func (o tax) GetName() string {
	return "TAX"
}

type txa struct{}

func (o txa) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.CPU.X
	console.CPU.A = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o txa) getVariations() []Variation {
	return []Variation{
		{opcode: 0x8A, addressingMode: nil, cycles: 2},
	}
}

func (o txa) GetName() string {
	return "TXA"
}

type tay struct{}

func (o tay) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.CPU.A
	console.CPU.Y = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o tay) getVariations() []Variation {
	return []Variation{
		{opcode: 0xA8, addressingMode: nil, cycles: 2},
	}
}

func (o tay) GetName() string {
	return "TAY"
}

type tya struct{}

func (o tya) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.CPU.Y
	console.CPU.A = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o tya) getVariations() []Variation {
	return []Variation{
		{opcode: 0x98, addressingMode: nil, cycles: 2},
	}
}

func (o tya) GetName() string {
	return "TYA"
}

type tsx struct{}

func (o tsx) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := byte(console.CPU.SP & 0x00FF)
	console.CPU.X = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o tsx) getVariations() []Variation {
	return []Variation{
		{opcode: 0xBA, addressingMode: nil, cycles: 2},
	}
}

func (o tsx) GetName() string {
	return "TSX"
}

type txs struct{}

func (o txs) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := int(console.CPU.X) | 0x0100
	console.CPU.SP = uint16(value)
	return variation.cycles
}

func (o txs) getVariations() []Variation {
	return []Variation{
		{opcode: 0x9A, addressingMode: nil, cycles: 2},
	}
}

func (o txs) GetName() string {
	return "TXS"
}

type pla struct{}

func (o pla) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.StackPopData()
	console.CPU.A = value
	console.CPU.SetZN(value)
	return variation.cycles
}

func (o pla) getVariations() []Variation {
	return []Variation{
		{opcode: 0x68, addressingMode: nil, cycles: 4},
	}
}

func (o pla) GetName() string {
	return "PLA"
}

type pha struct{}

func (o pha) Exec(console *processor.Console, variation *Variation, state *State) int {
	console.StackPushData(console.CPU.A)
	return variation.cycles
}

func (o pha) getVariations() []Variation {
	return []Variation{
		{opcode: 0x48, addressingMode: nil, cycles: 3},
	}
}

func (o pha) GetName() string {
	return "PHA"
}

type plp struct{}

func (o plp) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.StackPopData()
	flags := value&processor.NotBreakBit | processor.BFlag
	console.CPU.Flags = flags
	return variation.cycles
}

func (o plp) getVariations() []Variation {
	return []Variation{
		{opcode: 0x28, addressingMode: nil, cycles: 4},
	}
}

func (o plp) GetName() string {
	return "PLP"
}

type php struct{}

func (o php) Exec(console *processor.Console, variation *Variation, state *State) int {
	value := console.CPU.Flags | processor.BreakBit | processor.BFlag
	console.StackPushData(value)
	return variation.cycles
}

func (o php) getVariations() []Variation {
	return []Variation{
		{opcode: 0x08, addressingMode: nil, cycles: 3},
	}
}

func (o php) GetName() string {
	return "PHP"
}

var MoveOpCodes = []OpCode{
	lda{},
	sta{},
	ldx{},
	stx{},
	ldy{},
	sty{},
	tax{},
	txa{},
	tay{},
	tya{},
	tsx{},
	txs{},
	pla{},
	pha{},
	plp{},
	php{},
}
